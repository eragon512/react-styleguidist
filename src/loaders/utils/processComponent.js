const fs = require('fs');
const path = require('path');
const frontMatter = require('front-matter');
const getNameFromFilePath = require('./getNameFromFilePath');
const requireIt = require('./requireIt');
const slugger = require('./slugger');

const propsLoader = path.resolve(__dirname, '../props-loader.js');

/**
 * Loads the metadata from the markdown file of the Component.
 *
 * @param {string} filepath
 * @returns {object}
 */
function getComponentMetadata(filepath) {
	const componentMarkdownFilePath = filepath
		.substring(0, filepath.length - path.extname(filepath).length)
		.concat('.md');

	const componentReadmeMarkdownFilePath = path.join(path.dirname(filepath), 'Readme.md');

	let markdownFilePath;
	if (fs.existsSync(componentMarkdownFilePath)) {
		markdownFilePath = componentMarkdownFilePath;
	} else if (fs.existsSync(componentReadmeMarkdownFilePath)) {
		markdownFilePath = componentReadmeMarkdownFilePath;
	} else {
		return {};
	}

	const componentMarkdown = fs.readFileSync(markdownFilePath, 'utf8');
	if (!frontMatter.test(componentMarkdown)) {
		return {};
	}
	const componentMetadata = frontMatter(componentMarkdown).attributes;
	return componentMetadata;
}

/**
 * Return an object with all required for style guide information for a given component.
 *
 * @param {string} filepath
 * @param {object} config
 * @returns {object}
 */
module.exports = function processComponent(filepath, config) {
	const componentPath = path.relative(config.configDir, filepath);
	const componentName = getNameFromFilePath(filepath);
	const examplesFile = config.getExampleFilename(filepath);
	return {
		filepath: componentPath,
		slug: slugger.slug(componentName),
		pathLine: config.getComponentPathLine(componentPath),
		module: requireIt(filepath),
		props: requireIt(`!!${propsLoader}!${filepath}`),
		hasExamples: examplesFile && fs.existsSync(examplesFile),
		metadata: getComponentMetadata(filepath),
	};
};
