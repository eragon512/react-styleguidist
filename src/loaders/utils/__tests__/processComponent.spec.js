import path from 'path';
import processComponent from '../processComponent';

const config = {
	configDir: __dirname,
	getExampleFilename: componentpath => path.join(path.dirname(componentpath), 'Readme.md'),
	getComponentPathLine: componentpath => componentpath,
};

it('processComponent() should return an object for section with content', () => {
	const result = processComponent('pizza.js', config);

	expect(result).toMatchSnapshot();
});

it('processComponent() should return an object for section with content', () => {
	const result = processComponent('test/components/Button/Button.js', config);

	expect(result).toMatchSnapshot();
});

it('processComponent() should return an object with for section with content and metadata in markdown', () => {
	const result = processComponent('test/components/Placeholder/Placeholder.js', config);

	expect(result).toMatchSnapshot();
});
