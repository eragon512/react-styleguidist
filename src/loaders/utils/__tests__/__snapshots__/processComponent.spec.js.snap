// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processComponent() should return an object for section with content 1`] = `
Object {
  "filepath": "../../../../pizza.js",
  "hasExamples": true,
  "metadata": Object {},
  "module": Object {
    "require": "pizza.js",
  },
  "pathLine": "../../../../pizza.js",
  "props": Object {
    "require": "!!~/src/loaders/props-loader.js!pizza.js",
  },
  "slug": "pizza",
}
`;

exports[`processComponent() should return an object for section with content 2`] = `
Object {
  "filepath": "../../../../test/components/Button/Button.js",
  "hasExamples": true,
  "metadata": Object {},
  "module": Object {
    "require": "test/components/Button/Button.js",
  },
  "pathLine": "../../../../test/components/Button/Button.js",
  "props": Object {
    "require": "!!~/src/loaders/props-loader.js!test/components/Button/Button.js",
  },
  "slug": "button",
}
`;

exports[`processComponent() should return an object with for section with content and metadata in markdown 1`] = `
Object {
  "filepath": "../../../../test/components/Placeholder/Placeholder.js",
  "hasExamples": false,
  "metadata": Object {
    "author": "author",
    "title": "Placeholder",
  },
  "module": Object {
    "require": "test/components/Placeholder/Placeholder.js",
  },
  "pathLine": "../../../../test/components/Placeholder/Placeholder.js",
  "props": Object {
    "require": "!!~/src/loaders/props-loader.js!test/components/Placeholder/Placeholder.js",
  },
  "slug": "placeholder",
}
`;
